src/daemon/daemon.cpp

```diff
@@ -2668,6 +2680,82 @@ catch (const std::exception& e)
     status_promise->set_value(grpc::Status(grpc::StatusCode::INTERNAL, e.what(), ""));
 }
 
+void mp::Daemon::clone(const CloneRequest* request,
+                       grpc::ServerReaderWriterInterface<CloneReply, CloneRequest>* server,
+                       std::promise<grpc::Status>* status_promise)
+{
+    try
+    {
+        config->factory->require_clone_support();
+        mpl::ClientLogger<CloneReply, CloneRequest> logger{mpl::level_from(request->verbosity_level()),
+                                                           *config->logger,
+                                                           server};
+
+        const auto& source_name = request->source_name();
+        const auto [instance_trail, status] = find_instance_and_react(operative_instances,
+                                                                      deleted_instances,
+                                                                      source_name,
+                                                                      require_operative_instances_reaction);
+        if (status.ok())
+        {
+            const std::string destination_name = generate_destination_instance_name_for_clone(*request);
+
+            auto rollback_clean_up_all_resource_of_dest_instance = sg::make_scope_guard(
+                [this, destination_name]() noexcept -> void { release_resources(destination_name); });
+
+            assert(instance_trail.index() == 0);
+            const auto source_vm_ptr = std::get<0>(instance_trail)->second;
+            assert(source_vm_ptr);
+            const VirtualMachine::State source_vm_state = source_vm_ptr->current_state();
+            if (source_vm_state != VirtualMachine::State::stopped && source_vm_state != VirtualMachine::State::off)
+            {
+                throw std::runtime_error("Please stop instance " + source_name + " before you clone it.");
+            }
```

Comment:
This check can be moved to before creating the scope guard. Avoids some unnecessary work of trying to clean up an instance that doesn't exist yet.

It can even be moved before generating the clone name since checking if the instance is stopped comes first in the logical sequence of cloning a vm.

Reply:
Moving before `generate_destination_instance_name_for_clone` is debatable. That function has one check which checks the format validity of the destination name. I feel that one should come before checking if the src instance is stopped. 

Other generating name operation indeed should come after this check. So overall, it can be an improvement.